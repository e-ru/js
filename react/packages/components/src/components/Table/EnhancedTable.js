import React, { useState } from "react";
import PropTypes from "prop-types";

import { makeStyles } from "@material-ui/core/styles";
import Paper from "@material-ui/core/Paper";
import Table from "@material-ui/core/Table";
import TableBody from "@material-ui/core/TableBody";
import TableRow from "@material-ui/core/TableRow";
import TableCell from "@material-ui/core/TableCell";
import TablePagination from "@material-ui/core/TablePagination";
import LinearProgress from "@material-ui/core/LinearProgress";

import EnhancedTableToolbar from "./EnhancedTableToolbar";
import EnhancedTableHead from "./EnhancedTableHead";

import { stableSort, getSorting } from "../../utils/table";

const useStyles = makeStyles(theme => ({
  paper: {
    width: "100%",
    marginBottom: theme.spacing(2),
  },
  tableWrapper: {
    overflowX: "auto",
  },
}));

const EnhancedTable = ({ title, rows, headRows, tableRowComponent: TableRowComponent, inProgress }) => {
  const [selected, setSelected] = useState([]);
  const [order, setOrder] = useState("asc");
  const [orderBy, setOrderBy] = useState("username");
  const [page, setPage] = useState(0);
  const [rowsPerPage, setRowsPerPage] = useState(5);

  const dense = false;

  function handleSelectAllClick(event) {
    if (event.target.checked) {
      const newSelecteds = rows.map(row => row.name);
      setSelected(newSelecteds);
      return;
    }
    setSelected([]);
  }

  function handleRequestSort(event, property) {
    const isDesc = orderBy === property && order === "desc";
    setOrder(isDesc ? "asc" : "desc");
    setOrderBy(property);
  }

  const isSelected = name => selected.indexOf(name) !== -1;

  function handleClick(event, name) {
    const selectedIndex = selected.indexOf(name);
    let newSelected = [];

    if (selectedIndex === -1) {
      newSelected = newSelected.concat(selected, name);
    } else if (selectedIndex === 0) {
      newSelected = newSelected.concat(selected.slice(1));
    } else if (selectedIndex === selected.length - 1) {
      newSelected = newSelected.concat(selected.slice(0, -1));
    } else if (selectedIndex > 0) {
      newSelected = newSelected.concat(selected.slice(0, selectedIndex), selected.slice(selectedIndex + 1));
    }

    setSelected(newSelected);
  }

  function handleChangePage(event, newPage) {
    setPage(newPage);
  }

  function handleChangeRowsPerPage(event) {
    setRowsPerPage(+event.target.value);
  }

  const emptyRows = rowsPerPage - Math.min(rowsPerPage, rows.length - page * rowsPerPage);

  const height = document.documentElement.clientHeight - 64 - 48 - 64 - 71 - 1 - (inProgress ? 4 : 0);
  const classes = useStyles();
  return (
    <Paper className={classes.paper}>
      <EnhancedTableToolbar title={title} numSelected={selected.length} />
      <div className={classes.tableWrapper} style={{ height: `${height}px` }}>
        <Table className={classes.table} aria-labelledby="tableTitle" size={dense ? "small" : "medium"}>
          <EnhancedTableHead
            numSelected={selected.length}
            order={order}
            orderBy={orderBy}
            onSelectAllClick={handleSelectAllClick}
            onRequestSort={handleRequestSort}
            rowCount={rows.length}
            headRows={headRows}
          />
          {!inProgress && (
            <TableBody>
              {stableSort(rows, getSorting(order, orderBy))
                .slice(page * rowsPerPage, page * rowsPerPage + rowsPerPage)
                .map((row, index) => {
                  const isItemSelected = isSelected(row.name);
                  const labelId = `enhanced-table-checkbox-${index}`;
                  return (
                    <TableRowComponent
                      key={row.name}
                      row={row}
                      isItemSelected={isItemSelected}
                      labelId={labelId}
                      handleClick={handleClick}
                    />
                  );
                })}
              {emptyRows > 0 && (
                <TableRow style={{ height: 49 * emptyRows }}>
                  <TableCell colSpan={6} />
                </TableRow>
              )}
            </TableBody>
          )}
        </Table>
      </div>
      {inProgress && <LinearProgress />}
      <TablePagination
        rowsPerPageOptions={[5, 10, 25]}
        component="div"
        count={rows.length}
        rowsPerPage={rowsPerPage}
        page={page}
        backIconButtonProps={{
          "aria-label": "Previous Page",
        }}
        nextIconButtonProps={{
          "aria-label": "Next Page",
        }}
        onChangePage={handleChangePage}
        onChangeRowsPerPage={handleChangeRowsPerPage}
      />
    </Paper>
  );
};

EnhancedTable.defaultProps = {
  inProgress: false,
};

EnhancedTable.propTypes = {
  title: PropTypes.string.isRequired,
  rows: PropTypes.array.isRequired,
  headRows: PropTypes.array.isRequired,
  tableRowComponent: PropTypes.func.isRequired,
  inProgress: PropTypes.bool,
};

export default EnhancedTable;
