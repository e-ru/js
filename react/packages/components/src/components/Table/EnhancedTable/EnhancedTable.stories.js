import React from "react";
import { storiesOf } from "@storybook/react";

import EnhancedTable from "./EnhancedTable";

storiesOf("EnhancedTable", module).add("main item", () => (
  <EnhancedTable
    title="test title"
    rows={[]}
    headRows={[
      { id: "row1", numeric: false, disablePadding: true, label: "row1" },
      { id: "row2", numeric: false, disablePadding: false, label: "row2" },
    ]}
    tableRowComponent={() => console.log("onClickHandler clicked")}
  />
));
