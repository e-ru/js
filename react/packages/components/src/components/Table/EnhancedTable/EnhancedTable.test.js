import React from "react";
import { configure, shallow } from "enzyme";
import Adapter from "enzyme-adapter-react-16";

import EnhancedTable from "./EnhancedTable";

configure({ adapter: new Adapter() });

function setup() {
  const props = {
    title: "test title",
    rows: [],
    headRows: [
      { id: "row1", numeric: false, disablePadding: true, label: "row1" },
      { id: "row2", numeric: false, disablePadding: false, label: "row2" },
    ],
    tableRowComponent: jest.fn(),
  };

  const wrapper = shallow(<EnhancedTable {...props} />);

  return {
    wrapper,
  };
}

describe("<EnhancedTable />", () => {
  test("renders", () => {
    const { wrapper } = setup();
    expect(wrapper).toMatchSnapshot();
  });
});
