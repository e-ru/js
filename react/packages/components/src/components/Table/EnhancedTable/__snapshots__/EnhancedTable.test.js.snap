// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<EnhancedTable /> renders 1`] = `ShallowWrapper {}`;
